src/mpi/coll/allgather/allgather.c

```diff
@@ -170,57 +170,83 @@ int MPIR_Allgather_impl(const void *sendbuf, int sendcount, MPI_Datatype sendtyp
 
     if (comm_ptr->comm_kind == MPIR_COMM_KIND__INTRACOMM) {
         /* intracommunicator */
-        switch (MPIR_CVAR_ALLGATHER_INTRA_ALGORITHM) {
-            case MPIR_CVAR_ALLGATHER_INTRA_ALGORITHM_brucks:
-                mpi_errno =
-                    MPIR_Allgather_intra_brucks(sendbuf, sendcount, sendtype, recvbuf, recvcount,
-                                                recvtype, comm_ptr, errflag);
-                break;
-            case MPIR_CVAR_ALLGATHER_INTRA_ALGORITHM_recursive_doubling:
-                mpi_errno =
-                    MPIR_Allgather_intra_recursive_doubling(sendbuf, sendcount, sendtype, recvbuf,
-                                                            recvcount, recvtype, comm_ptr, errflag);
-                break;
-            case MPIR_CVAR_ALLGATHER_INTRA_ALGORITHM_ring:
-                mpi_errno =
-                    MPIR_Allgather_intra_ring(sendbuf, sendcount, sendtype, recvbuf, recvcount,
-                                              recvtype, comm_ptr, errflag);
-                break;
-            case MPIR_CVAR_ALLGATHER_INTRA_ALGORITHM_nb:
-                mpi_errno =
-                    MPIR_Allgather_allcomm_nb(sendbuf, sendcount, sendtype, recvbuf, recvcount,
-                                              recvtype, comm_ptr, errflag);
-                break;
-            case MPIR_CVAR_ALLGATHER_INTRA_ALGORITHM_auto:
-                MPL_FALLTHROUGH;
-            default:
-                mpi_errno =
-                    MPIR_Allgather_intra_auto(sendbuf, sendcount, sendtype, recvbuf, recvcount,
-                                              recvtype, comm_ptr, errflag);
-                break;
+        int i;
+        for (i = 0; MPIR_CVAR_ALLGATHER_INTRA_ALGORITHM[i] != MPIR_CVAR_ENUM_NULL; i++) {
```

Comment:
If I set `MPIR_CVAR_ALLGATHER_INTRA_ALGORITHM=brucks,recursive_doubling`, what will happen?

Reply:
In this case, it'll try to use the `brucks` algorithm first.  If it cannot, because `brucks` has some limitations (e.g., if it only works when the number of processes is prime), then it'll fallback to `recursive_doubling`.  If `recursive_doubling` cannot work because it has some other limitations (e.g., if it only works when the message size is a power of 17), then it'll fallback to the original logic, which is: (1) abort, (2) print something and abort, or (3) silently fallback to the auto/json algorithm, depending on a different CVAR setting.

Reply:
I understand that's what you're going for, but I don't see where this code goes to the second option. It looks to me like after you call `MPIR_Allgather_intra_brucks`, it's going to `break` out of the `switch` and then `break` again out of the `for` loop to go to `fallback` and call the `auto` version of the function. It seems like you need to check `mpi_errno` to see if the function was successful or not and `continue` if not.

Reply:
case (2) should be print something and fallback

Reply:
Does it really matter which cases prints vs. which case aborts?  The README.envvar mentions it anyway.

Reply:
We need a case which prints (if algo is not supported) but continues running, instead of aborting. So case(2) can be "print and fallback" instead of "print and abort".

Reply:
Oh, I understand @zhenggb72's comment now.  Yes, we print and fallback, not print and abort.  I misspoke.